### DISCUSSION TITLE
Getting started.

### DISCUSSION THREAD

**Discussion by arthurwolf** [ggml-org/llama.cpp#2597] [Q&A] [answered]
Hey!

My project currently uses GPT3.5 with API calls (which ends up being pretty expensive), and testing out the demo for llama2-70b-chat, it looks like it'd work well enough for at least part of the prompts I'm using.

But things seem to be moving pretty fast, and I'm not clear on exactly what is possible and what is the best path forward.

I have:
* A CPU with 12 cores and about 40GB of free RAM
* A GPU (3070) with about 10GB of free VRAM (could get a bit more by turning screens off).

I can get if needed:
* A server with any amount of RAM, up to 256GB if needed, and a beefy CPU.

I want:
* To run llama2-70b-chat for my code to query/call with prompts.

My question is: what's my best path forward?

I'm been able to get the 7b version running following the instructions, but my local computer doesn't have enough RAM to run the 70b version (thus the mention of a server above, which I'm getting ready to start renting).

I'm not sure exactly how to run on the GPU, and if I have enough RAM to do that or not. I'm not sure I follow the instructions for that.

I'm also not sure what my options are besides llama.cpp, if any. I found this project, but I'm not sure if there are any others out there I could try / that would be a better option for what I'm trying to do here.

Any help/pointers would be very much appreciated.

Cheers.

**Comment by ianscrivener** [comment, marked as answer]
@arthurwolf,
llama.cpp can definately do the job! eg "I'm succesfully running llama-2-70b-chat.ggmlv3.q3_K_S on my 32 GB RAM on cpu with speed of 1.2 tokens/s without any GPU offloading (i dont have a descrete gpu), using full 4k context and kobold.cpp on windows 11 pro. [mentioned here](https://huggingface.co/TheBloke/Llama-2-70B-Chat-GGML/discussions/1)

Were it me; I'd start my experimentations with 38.8Gb [llama-2-70b-chat.ggmlv3.q4_0.bin](https://huggingface.co/TheBloke/Llama-2-70B-Chat-GGML/resolve/main/llama-2-70b-chat.ggmlv3.q4_0.bin)

We know GPU is significantly faster than CPU. I use Runpod for Cloud GPU. An A6000 has 62Gb GPU RAM... so could run the above Llama-2-70B-Chat model in GPU. RunPod has a template, also by TheBloke, that is a good starting point [doco here](https://github.com/TheBlokeAI/dockerLLM/blob/main/README_Runpod_LocalLLMsUI.md)

There seems to be far more disussion of the MLOps stuff (model selection, hardware specs, setups etc) on [Reddit](https://www.reddit.com/r/LocalLLaMA/) and Huggingspace than here.

[The Bloke has a Discord channel](https://discord.com/invite/Jq4vkcDakD) as well - personally I'd start there.

**Reply by arthurwolf** [reply]
Thanks a megaton for the detailed reply, that's going to help immensely.

**Reply by arthurwolf** [reply]
So I'm doing 

```
./server -m ~/models/llama-2-70b.ggmlv3.q5_0.bin -gqa 8 --host x.x.x.x --port 8123
```

and I'm getting 

```
llama_print_timings:        load time = 159370.45 ms
llama_print_timings:      sample time =    64.06 ms /    20 runs   (    3.20 ms per token,   312.18 tokens per second)
llama_print_timings: prompt eval time = 159365.63 ms /   341 tokens (  467.35 ms per token,     2.14 tokens per second)
llama_print_timings:        eval time = 59522.49 ms /    19 runs   ( 3132.76 ms per token,     0.32 tokens per second)
llama_print_timings:       total time = 218965.89 ms
```

That seems low/slow compared to what I've seen around. This is with a 24 core AMD CPU, no GPU offloading.

Is there anything I can tune/play with that would change speed significantly without impacting quality?

**Reply by ianscrivener** [reply]
you are running the model in CPU. Try adding `--n_gpu_layers 1000` to use GPU

**Reply by arthurwolf** [reply]
I'm unfortunately unable to use the GPU, I've spent a full day installing/reinstalling CUDA and installing/reinstalling Nvidia drivers, and I can absolutely never get it to build, it always complains about something, most of the time version mismatches. I really wish there was a straightforward way to set this up...

**Reply by ianscrivener** [reply]
Have you tried a running llama.cpp inside a Docker container? That will side step some of the version issues.

1) check your base/host OS nvidia drivers with `nvidia-smi`
2) Install NVIDIA Container Toolkit to your host. "This integrates into Docker Engine to automatically configure your containers for GPU support"
3) the [llama.cpp dockerfile is here](https://github.com/ggerganov/llama.cpp/blob/master/.devops/full-cuda.Dockerfile) 

Then _something_ like
```
git clone https://github.com/ggerganov/llama.cpp.git
cd .devops
docker build --file main-cuda.Dockerfile

docker run --gpus all ...
```

**Comment by jasonkaplan79** [comment]
These "Getting started" instructions no longer seem to be accurate, at least, they dont work from a Mac Terminal.

I tried this command:
```
./server -m ~/models/llama-2-70b.ggmlv3.q5_0.bin -gqa 8 --host x.x.x.x --port 8123

```

![image](https://github.com/user-attachments/assets/ad497c99-05ac-4f66-8682-b069dfa673c2)

I started another discussion here for beginners, hoping that it will help me (and others down the line): https://github.com/ggerganov/llama.cpp/discussions/10631
